components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/jwks/JwksEndpoint.java

```diff
@@ -65,50 +68,84 @@ public String jwks() {
+                    Certificate cert = keystore.getCertificate(alias);
+                    certKeyPair.put(alias,cert);
+                }
             }
-            return buildResponse(tenantDomain, tenantId, publicKey);
+            return buildResponse(certKeyPair);
         } catch (Exception e) {
-            String errorMessage = "Error while generating the keyset for " + tenantDomain + " tenant domain.";
+            String errorMessage = "Error while generating the keyset for tenant domain: " + tenantDomain;
             return logAndReturnError(errorMessage, e);
         }
     }
 
-    private String buildResponse(String tenantDomain, int tenantId, RSAPublicKey publicKey)
+    private String buildResponse(HashMap<String, Certificate> certificates)
             throws IdentityOAuth2Exception, ParseException {
+
         JSONArray jwksArray = new JSONArray();
         JSONObject jwksJson = new JSONObject();
-
-        RSAKey.Builder jwk = new RSAKey.Builder(publicKey);
-        jwk.keyID(OAuth2Util.getThumbPrint(tenantDomain, tenantId));
-        jwk.algorithm(JWSAlgorithm.RS256);
-        jwk.keyUse(KeyUse.parse(KEY_USE));
-        jwksArray.put(jwk.build().toJSONObject());
+        OAuthServerConfiguration config = OAuthServerConfiguration.getInstance();
+        JWSAlgorithm accessTokenSignAlgorithm =
+                OAuth2Util.mapSignatureAlgorithmForJWSAlgorithm(config.getIdTokenSignatureAlgorithm());
+        for (Map.Entry certKeyPair : certificates.entrySet()) {
+            Certificate cert = (Certificate) certKeyPair.getValue();
+            String alias = (String) certKeyPair.getKey();
+            RSAPublicKey publicKey = (RSAPublicKey) cert.getPublicKey();
+            RSAKey.Builder jwk = new RSAKey.Builder(publicKey);
+            jwk.keyID(OAuth2Util.getThumbPrint(cert, alias));
+            jwk.algorithm(accessTokenSignAlgorithm);
+            jwk.keyUse(KeyUse.parse(KEY_USE));
+            jwksArray.put(jwk.build().toJSONObject());
+        }
+        //read all the signature algorithms from identity.xml and put them in a list if there are different
+        ArrayList<JWSAlgorithm> algorithms = new ArrayList<>();
+        algorithms.add(accessTokenSignAlgorithm);
+        JWSAlgorithm idTokenSignAlgorithm =
+                OAuth2Util.mapSignatureAlgorithmForJWSAlgorithm(config.getIdTokenSignatureAlgorithm());
+        if (!accessTokenSignAlgorithm.equals(idTokenSignAlgorithm)) {
+            algorithms.add(idTokenSignAlgorithm);
+        }
+        JWSAlgorithm userInfoSignAlgorithm =
+                OAuth2Util.mapSignatureAlgorithmForJWSAlgorithm(config.getUserInfoJWTSignatureAlgorithm());
+        if (!accessTokenSignAlgorithm.equals(userInfoSignAlgorithm)
+                && !idTokenSignAlgorithm.equals(userInfoSignAlgorithm)) {
+            algorithms.add(userInfoSignAlgorithm);
+        }
+        //create the JWKS for different algorithms
+        for (Map.Entry certKeyPair : certificates.entrySet()) {
```

Comment:
can we refactor the code so that you iterate over certificates only one time? Check the possibility.

Reply:
okay